klippy/extras/safe_power.py

```diff
@@ -0,0 +1,116 @@
+# Stepper Driver Anti-SNAFU Protection support
+#
+# Copyright (C) 2021  Lars R. Hansen <popshansen@hotmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import logging, time
+
+class BackwardError(Exception):
+    pass
+
+SF_INVERT_STEP = 1<<2
+
+class Safe_Power:
+    error = BackwardError
+    def __init__(self, config):
+        self.name = config.get_name()
+        self.printer = config.get_printer()
+        self.reactor = self.printer.get_reactor()
+        ppins = self.printer.lookup_object('pins')
+        pin_params = ppins.lookup_pin(config.get('pin'))
+        self.mcu = pin_params['chip']
+        self.oid = self.mcu.create_oid()
+        self.pin = pin_params['pin']
+        self.mcu_pin = self.mcu.setup_pin('digital_out', pin_params)
+        self.mcu_pin.setup_start_value(0, 0, False)
+        self.set_digital_out_cmd = self.read_digital_in_cmd = None
+        self.steppers = {}
+        for s in config.get_prefix_sections('stepper_'):
+            self._register_stepper(s, ppins)
+        for s in config.get_prefix_sections('extruder'):
+            self._register_stepper(s, ppins)
+        for s in config.get_prefix_sections('manual_stepper'):
+            self._register_stepper(s, ppins)
+        if config.has_section('dual_carriage'):
+            self._register_stepper(config.getsection('dual_carriage'), ppins)
+        #Check config order
+        for name, obj in self.printer.lookup_objects():
+            if 'tmc' in name.lower():
```

Comment:
A search for 'tmc' may conflict with arbitrary user names.  Consider using `name.startswith("tmc")`.